Implement HEAD for /upload

laanwj opened the pull request:
Factor out the checking if uploading is allowed to a function `upload_blob_prechecks`, use this in the `PUT` handler, as well as the new `HEAD` handler `upload_head_handler`.

Closes #9. Fixes the current compatibility issue with nostrudel-next.

laanwj commented:
Changing to draft for now. Although it makes nostrudel work, this doesn't entirely implement the behavior described in [BUD-06](https://github.com/hzrd149/blossom/blob/master/buds/06.md#head-upload---upload-requirements).

laanwj commented:
Better now
```
$ curl --head http://127.0.0.1:3030/upload
HTTP/1.1 401 Unauthorized
x-reason: Authorization event required to upload a blob
content-length: 0
vary: origin, access-control-request-method, access-control-request-headers
access-control-allow-origin: *
date: Sat, 16 Nov 2024 21:23:35 GMT
```

0xtrr commented:
> i'm not sure we can do much checking, except for length against the allowed size.

We can check the size and mimetype. We'd have to have a sha blacklist to be able to validate the sha header. If anyone needs that, we can add it later on imo.

laanwj commented:
> We can check the size and mimetype. We'd have to have a sha blacklist to be able to validate the sha header. If anyone needs that, we can add it later on imo.

Thanks. i've added the check for the content length header, and moved the check for content type to the precheck function as well.

0xtrr commented on src/handlers/mod.rs:
```diff
@@ -338,41 +342,87 @@ pub async fn list_blobs_handler(
+    let blob_size_in_mb = bytes_to_mb(content_length as f64);
+    if blob_size_in_mb > app_state.config.upload.max_size {
+        return Err((StatusCode::BAD_REQUEST, format!(
+                "Blob size is {} MB, max upload size is {}",
```
nitpick: Should we add "MB" behind the max upload size as well?

laanwj replied:
Yes that makes sense, will add

0xtrr reviewed (approved):
Approved but consider my little nitpick question/proposal on the content length error message.

0xtrr commented:
Actually, I see that tests are failing after merging this because we now require the PUT /upload request to always have the content-length request header. BUD-02 specifies that we MAY use the header for checking data size but it's not strictly required. We should probably check only if it's there? Else, we rely on the size check on the actual blob bytes. What do you think @laanwj?

0xtrr commented:
I added a GitHub workflow to the project which will run on PRs targeting main and on any commit to main. I'll write a bunch of unit tests as well to help us ensure correct behavior and future maintenance.

laanwj commented:
> We should probably check only if it's there? 

Yes, i think it's fine to ignore the check if the header is missing, it seems like an advisory thing.

Sorry for not running the tests. Would make sense to run them on PRs too.
